pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,278 @@
+	Message string
+}
+
+// FlowSchemaConditionType is a valid value for FlowSchemaStatusCondition.Type
+type FlowSchemaConditionType string
+
+// FlowSchemaConditionStatus is a valid value for FlowSchemaStatusCondition.Status
+type FlowSchemaConditionStatus string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec
+	// Current status of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+	// Items is a list of request-priorities.
+	Items []PriorityLevelConfiguration
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+type PriorityLevelConfigurationSpec struct {
+	// GlobalDefault specifies whether this priority level should be considered as the default priority for requests
+	// that do not have any priority level. Only one PriorityClass can be marked as `globalDefault`.
+	// +optional
+	GlobalDefault bool
+	// Exempt defines whether the priority level is exempted or not.  There should be at most one exempt priority level.
+	// Being exempt means that requests of that priority are not subject to concurrency limits (and thus are never queued)
+	// and do not detract from the concurrency available for non-exempt requests. In a more sophisticated system, the
+	// exempt priority level would be the highest priority level. The field is default to false and only those system
+	// preset priority level can be exempt.
+	// +optional
+	Exempt bool
+	// AssuredConcurrencyShares is a positive number for a non-exempt priority level, representing the weight by which
+	// the priority level shares the concurrency from the global limit. The concurrency limit of an apiserver is divided
+	// among the non-exempt priority levels in proportion to their assured concurrency shares. Basically this produces
+	// the assured concurrency value (ACV) for each priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+	//
```

Comment:
I suggest adding something like this: throughput not used is available for other priority levels to use. Set this value to indicate how much throughput this priority level can consume when the system is completely loaded.

Reply:
I thought we agreed that the first version would not do that.  The KEP currently says that.

Reply:
Where does it say that in the KEP?

It might be the jet lag talking, but without this property it seems like a configuration that isn't well matched to the types of requests an apiserver actually receives is likely to drastically reduce the throughput of the system, i.e., enabling it by default seems very risky.

If I agreed to start like this I think I must have intended to mean as a way of getting started; I'm willing to be convinced otherwise but I currently think we need that property before this can be beta if not alpha?

Reply:
The KEP says this in the first paragraph of the Proposal, at https://github.com/kubernetes/enhancements/blob/7a144ac413b0d9b3fbe43c0bc5dd843e2e7c19d7/keps/sig-api-machinery/20190228-priority-and-fairness.md#proposal

This is reiterated in https://github.com/kubernetes/enhancements/blob/7a144ac413b0d9b3fbe43c0bc5dd843e2e7c19d7/keps/sig-api-machinery/20190228-priority-and-fairness.md#dispatching

Reply:
And yes, I expect this first version is going to be unsatisfying --- and quite possibly very annoying.  My hope is that it will provide a useful common baseline of experience and understanding.

Reply:
i remember we agreed that no borrowing in the alpha stage even though that's also what we literally want in a bigger picture of this feature

Reply:
After I thought about it more, I think this should still behave OK, since the server won't sit idle unless no queue has anything to do. We can probably go to beta with this and get experience.